core/ir/synthetic/encode.c

```diff
@@ -0,0 +1,91 @@
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+/* Copyright (c) 2003-2007 Determina Corp. */
+/* Copyright (c) 2001-2003 Massachusetts Institute of Technology */
+/* Copyright (c) 2001 Hewlett-Packard Company */
+
+/* encode.c -- an encoder for DR synthetic IR */
+
+#include "encode.h"
+
+#include "../globals.h"
+
+#define CATEGORY_BITS 22
+#define FLAGS_BITS 2
+#define NUM_OPND_BITS 4
+
+void
+encode_to_synth(dcontext_t *dcontext, instr_t *instr, byte *encoded_instr)
+{
+    uint encoding = 0;
+    uint shift = 0;
+
+    /*
+     * Encode category as synthetic opcode.
+     */
+    uint category = instr_get_category(instr);
+    encoding |= category;
+    shift += CATEGORY_BITS;
+
+    /*
+     * Encode flags.
+     */
+    // TOFIX: this is probably not the right way to look at arithmetic flags of an instr_t
+    uint flags = (uint)instr_arith_flags_valid(instr);
+    encoding |= (flags << shift);
+    shift += FLAGS_BITS;
+
+    /*
+     * Encode number of source operands.
+     */
+    uint num_srcs = (uint)instr_num_srcs(instr);
+    encoding |= (num_srcs << shift);
+    shift += NUM_OPND_BITS;
+
+    /*
+     * Encode number of destination operands.
+     */
+    uint num_dsts = (uint)instr_num_dsts(instr);
+    encoding |= (num_dsts << shift);
+
+    /*
+     * TODO: Encode registers.
+     */
+
+    /*
+     * Copy result to output encoded_instr.
+     */
+    memcpy(encoded_instr, &encoding, sizeof(uint));
```

Comment:
`sizeof(encoded_instr)` is safer in case its type gets changed.

Reply:
Since we want to copy only 4 bytes, I used `sizeof(encoding)` as before (`sizeof(encoded_instr)` would be 8 bytes, being a pointer).
I assume `encoded_instr` has enough space.